benchmarks/nemo/nemolite2d/manual_versions/psykal_opencl/time_step_mod.f90

```diff
@@ -69,860 +71,808 @@ SUBROUTINE invoke_time_step(ssha_t, sshn_t, sshn_u, sshn_v, hu, hv, un, &
-        CALL check_status('clSetKernelArg: arg 14 of momentum_u_code', ierr)
-        ierr = clSetKernelArg(kernel_obj, 15, C_SIZEOF(dy_t), C_LOC(dy_t))
-        CALL check_status('clSetKernelArg: arg 15 of momentum_u_code', ierr)
-        ierr = clSetKernelArg(kernel_obj, 16, C_SIZEOF(area_u), C_LOC(area_u))
-        CALL check_status('clSetKernelArg: arg 16 of momentum_u_code', ierr)
-        ierr = clSetKernelArg(kernel_obj, 17, C_SIZEOF(gphiu), C_LOC(gphiu))
-        CALL check_status('clSetKernelArg: arg 17 of momentum_u_code', ierr)
-        ierr = clSetKernelArg(kernel_obj, 18, C_SIZEOF(omega), C_LOC(omega))
-        CALL check_status('clSetKernelArg: arg 18 of momentum_u_code', ierr)
-        ierr = clSetKernelArg(kernel_obj, 19, C_SIZEOF(d2r), C_LOC(d2r))
-        CALL check_status('clSetKernelArg: arg 19 of momentum_u_code', ierr)
-        ierr = clSetKernelArg(kernel_obj, 20, C_SIZEOF(g), C_LOC(g))
-        CALL check_status('clSetKernelArg: arg 20 of momentum_u_code', ierr)
-        ierr = clSetKernelArg(kernel_obj, 21, C_SIZEOF(rdt), C_LOC(rdt))
-        CALL check_status('clSetKernelArg: arg 21 of momentum_u_code', ierr)
-        ierr = clSetKernelArg(kernel_obj, 22, C_SIZEOF(cbfr), C_LOC(cbfr))
-        CALL check_status('clSetKernelArg: arg 22 of momentum_u_code', ierr)
-        ierr = clSetKernelArg(kernel_obj, 23, C_SIZEOF(visc), C_LOC(visc))
-        CALL check_status('clSetKernelArg: arg 23 of momentum_u_code', ierr)
-    END SUBROUTINE momentum_u_code_set_args
+        use kind_params_mod, only: go_wp
+        USE clfortran
+        USE fortcl, ONLY: get_cmd_queues
+        real(go_wp), dimension(:,:), intent(in), target :: from
+        type(c_ptr), intent(in) :: to
+        integer, intent(in) :: startx, starty, nx, ny
+        logical, intent(in) :: blocking
+        integer(c_intptr_t) :: cl_mem
+        INTEGER(c_size_t) :: size_in_bytes, offset_in_bytes
+        INTEGER(c_intptr_t), pointer :: cmd_queues(:)
+        integer :: ierr
+        ! Copy complete ny rows (regardless of nx)
+        size_in_bytes = int(size(from, 1) * ny, 8) * c_sizeof(from(1,1))
+        offset_in_bytes = int(size(from, 1) * (starty - 1)) * c_sizeof(from(1,1))
+        cl_mem = transfer(to, cl_mem)
+        cmd_queues => get_cmd_queues()
+        ierr = clEnqueueWriteBuffer(cmd_queues(1), cl_mem, &
+            CL_TRUE, offset_in_bytes, size_in_bytes, C_LOC(from), 0, &
+            C_NULL_PTR, C_NULL_PTR)
+        CALL check_status('clEnqueueWriteBuffer', ierr)
+    end subroutine write_opencl
 
-    SUBROUTINE momentum_v_code_set_args(kernel_obj, va, un, vn, hu, hv, ht, &
-            ssha_v, sshn_t, sshn_u, sshn_v, tmask, dx_v, dx_t, &
-            dy_u, dy_v, dy_t, area_v, gphiv, omega, d2r, g, rdt, cbfr, visc)
-        USE clfortran, ONLY: clSetKernelArg
-        USE iso_c_binding, ONLY: c_sizeof, c_loc, c_intptr_t
+    subroutine initialise_device_grid(field)
+        USE fortcl, ONLY: create_rw_buffer
```

Comment:
Please add a comment describing what this routine does.

Reply:
Done